### ISSUE TITLE
Sub-domain boundary artifacts

### ISSUE THREAD

**Issue by sbrus89** [MPAS-Dev/MPAS-Model#140]
On the ocean/develop branch, solution artifacts appear at sub-domain boundaries in certain fields (density, kineticEnergyCell, salinity, temperature, relativeVorticityCell) as shown here 

![soma_default_kineticenergy_subdomain](https://user-images.githubusercontent.com/3514023/51696015-acd06b00-1fc1-11e9-87b5-4bf52ac732f8.png)

Reproducibility information:
  

- commit: 575586ebc
- Modules: 1) python/anaconda-2.7-climate   2) gcc/5.3.0   3) openmpi/1.10.5   4) netcdf/4.4.1   5) parallel-netcdf/1.5.0   6) pio/1.7.2
- make line: make clean gfortran CORE=ocean AUTOCLEAN=true
- run line: srun -n 4 ocean_model -n namelist.ocean -s streams.ocean
- test case: -o ocean -c soma -r 32km -t default
- uname -a: Linux gr-fe1.lanl.gov 3.10.0-862.14.4.1chaos.ch6.x86_64 #1 SMP Wed Sep 26 12:27:08 PDT 2018 x86_64 x86_64 x86_64 GNU/Linux

**Comment by pwolfram** [comment]
Thanks @sbrus89, @mark-petersen, @xylar, @vanroekel have any of you seen these types of issues with interprocessor boundaries?  This seems like something we need to resolve because it will obviously affect all our simulations.

**Comment by xylar** [comment]
It does seem surprising if this issue is specific to SOMA.  At the same time, it'd be surprising if this is a general issue and we just haven't noticed it in other simulations.  It would be worth finding out if this is due to a recent change.

It has all the hallmarks of a bad halo update, but hard to be sure.

**Comment by mark-petersen** [comment]
This is disturbing.  We fixed a bunch of these types of bugs very early on, and have a bfb test across two different partitions on the QU240 global in the nightly test suite.  But of course that test has a number of different settings from your tests above.

@sbrus89, as a simple test, could you turn off the time-varying forcing in your global hurricane case, and look at the ssh at that same time? Could you provide more details about this last case, if possible - COMPASS command and namelist alterations?

**Comment by pwolfram** [comment]
@mark-petersen, we originally saw these types of effects in a wetting / drying case for the Delaware and I thought they were related to something weird I was doing there but they appear to be persistent across a variety of cases and this is on vanilla ocean/develop, which is the disconcerting thing.

**Comment by pwolfram** [comment]
@mark-petersen, do you have any ideas on whether we can distinguish between whether this is a stand-alone or E3SM bug?  It is clearly happening in standalone and I'm concerned it is happening in E3SM too.

**Comment by vanroekel** [comment]
I took a look at SSH snapshots from a 30to10 G-case with E3SM master from 2 weeks ago and see no visible processor boundaries.

**Comment by mark-petersen** [comment]
Any configuration where partition boundaries are visible would fail E3SM PEM tests (exact comparison across two runs with different partition numbers).  We test a number of those in E3SM nightly regression suite on next and master.  We have a similar test  that passes QU240 on MPAS-stand alone nightly regression test.  I'll take a closer look at this particular case today or tomorrow.

**Comment by xylar** [comment]
@sbrus89, I'm hoping to see if I can reproduce the original problem on Cori.  Could you tell me how long you ran the SOMA 32km test case for?  The default configuration is for 10 minute time steps but a 1 minute 30 second run duration (and output interval).  Does that mean you ran for only 1 ten-minute time step?  Or did you modify the defaults?

**Comment by sbrus89** [comment]
@xylar, good point. I forgot to mention that I ran it for 1 day with the default 10 minute timestep.

**Comment by xylar** [comment]
Running SOMA 32km for one time step, I'm able to reproduce the issue with KE and vorticity but not with density, salinity and temperature.  Here are 2 examples.

KE:
![ke](https://user-images.githubusercontent.com/4179064/51923584-8eb4a180-23eb-11e9-9a44-50b2f7c59559.png)

salinity:
![salinity](https://user-images.githubusercontent.com/4179064/51923597-95431900-23eb-11e9-85a1-e56c6accb70c.png)

**Comment by sbrus89** [comment]
Could there be an issue with distributing the initial conditions, and that these perturbations contaminate other fields early on and then damp out over time?

**Comment by xylar** [comment]
If true, that should still fail the tests @mark-petersen mentioned.

**Comment by pwolfram** [comment]
@xylar, was ssh ok in the case above, correct?  If so, this implies corruption of the prognostic KE that feeds back via momentum advection.

**Comment by xylar** [comment]
SSH looks okay:
![ssh](https://user-images.githubusercontent.com/4179064/51935979-79e60700-2407-11e9-839d-a9c4a93c39b0.png)

**Comment by sbrus89** [comment]
@mark-petersen  Please ignore the global ocean images, @pwolfram and I were able to sort that out. I was looping over nCellsSolve instead of nCells when computing the wind stress from the time-varying wind speed (facepalm). 

However, that doesn't effect the default SOMA case, there is still an issue there.

**Comment by mark-petersen** [comment]
I was able to reproduce the error on the current head of `ocean/develop` 6ac5486e using
`-o ocean -c soma -r 32km -t default` with gnu on grizzly.
I ran that test case with 4 and 8 blocks, can see KE difference after one step:
```
 kineticEnergyCellAvg = 0, 1.57470975385003e-07
 kineticEnergyCellAvg = 0, 1.56589438918034e-07
```
here: `/lustre/scratch4/turquoise/mpeterse/runs/190129_test_soma_block_bfb/ocean/soma/32km/default`

**Comment by mark-petersen** [comment]
<details>
<summary> Here is the cvmix namelist: </summary> <p>

```
&cvmix
    config_use_cvmix = .true.
    config_cvmix_prandtl_number = 1.0
    config_use_cvmix_background = .true.
    config_cvmix_background_diffusion = 5.0e-6
    config_cvmix_background_viscosity = 1.0e-4
    config_use_cvmix_convection = .true.
    config_cvmix_convective_diffusion = 1.0
    config_cvmix_convective_viscosity = 1.0
    config_cvmix_convective_basedOnBVF = .true.
    config_cvmix_convective_triggerBVF = 0.0
    config_use_cvmix_shear = .true.
    config_cvmix_num_ri_smooth_loops = 2
    config_cvmix_use_BLD_smoothing = .true.
    config_cvmix_shear_mixing_scheme = 'KPP'
    config_cvmix_shear_PP_nu_zero = 0.005
    config_cvmix_shear_PP_alpha = 5.0
    config_cvmix_shear_PP_exp = 2.0
    config_cvmix_shear_KPP_nu_zero = 0.005
    config_cvmix_shear_KPP_Ri_zero = 0.7
    config_cvmix_shear_KPP_exp = 3
    config_use_cvmix_tidal_mixing = .false.
    config_use_cvmix_double_diffusion = .false.
    config_use_cvmix_kpp = .true.
    config_use_cvmix_fixed_boundary_layer = .false.
    config_cvmix_kpp_boundary_layer_depth = 30.0
    config_cvmix_kpp_criticalBulkRichardsonNumber = 0.25
    config_cvmix_kpp_matching = 'SimpleShapes'
    config_cvmix_kpp_EkmanOBL = .false.
    config_cvmix_kpp_MonObOBL = .false.
    config_cvmix_kpp_interpolationOMLType = 'quadratic'
    config_cvmix_kpp_surface_layer_extent = 0.1
    config_cvmix_kpp_surface_layer_averaging = 5.0
    configure_cvmix_kpp_minimum_OBL_under_sea_ice = 10.0
    config_cvmix_kpp_stop_OBL_search = 100.0
    config_cvmix_kpp_use_enhanced_diff = .true.
    config_cvmix_kpp_nonlocal_with_implicit_mix = .false.
/
```

</p>
</details>

**Comment by vanroekel** [comment]
@mark-petersen this appears to be a bug in rich_mix.  The coefficients are computed to nCellsSolve/nEdgesSolve, but the new cvmix interface requires coefficients to 1-halo, so unless there is a halo update in here there will be issues.  But I wholeheartedly agree with @xylar that we should never use anything but cvmix, at least until I finish the new version.  I also support removing all mixing options but cvmix.

**Comment by xylar** [comment]
@sbrus89 and @vanroekel, was this ever addressed?
